### PR TITLE
Session settings

### LINKED ISSUE #5534: [BUG] unable to choose different session storage methods, filesystem hard-coded
In the SessionServiceProvider, when gathering options for the Session classes in `registerOptions(Application $app)`, the method ends by adding `$app['session.options']`, effectively hard-coding these values and not allowing me to override them in any way. By switching this line and the ones adding `$app['session.storage.options']` (lines 168-170), we can overwrite the option values in our `.bolt.php` files and customize the way sessions are handled. (PR incoming)

### PR DESCRIPTION
**Pull request by dwolfhub**
First, as an implementer of Bolt, I would expect Bolt would use my default session settings set in my `php.ini` file. If for some reason I was using a crappy hosting service or was held back by a grumpy sysadmin, this PR would also let me override those session settings in my `config.yml` file.

It fixes #5534 for me because I wouldn't need to extend `Bolt/Application` to switch away from storing all sessions in `app/cache/.sessions/`. I could just make changes in my `php.ini` file or in my `config.yml` file.

### CODE REVIEW

**Comment by CarsonF** [comment]
I get what you are going for, but this is the wrong way. We already pull ini settings [here](https://github.com/bolt/bolt/blob/master/src/Provider/SessionServiceProvider.php#L164).

Instead we could add a config parameter for that. And `null` would use default value.

**Comment by dwolfhub** [comment]
@CarsonF you pull the ini settings on line 162, but then immediately override them with defaults from 'session.storage.options' on line 164. There is currently no way to use your default ini settings. In fact, there is currently no way to use anything other than `filesystem` with `cache://.sesssions`. It's all getting overridden. I'll take another crack at a PR.

**Comment by CarsonF** [comment]
Right so maybe we need to filter null values out of the `session.options` before it's `->add` to the defaults.

**Comment by dwolfhub** [comment]
@CarsonF by removing the configuration from the SessionProvider, i was able to not only decouple it like you wanted, but also fix my issue of not being able to configure the sessions without extending the bolt application. Please have a look. I think we're close.

File: src/Application.php [hunk 1]
```diff
@@ -84,6 +84,15 @@ protected function initSession()
...
+
+        $this['session.options'] = array_merge(
+            $this['session.options'],
+            $this['config']->get('general/session', [])
+        );
+        $this['session.options']['cookie_lifetime'] = $this['config']->get('general/cookies_lifetime');
+        $this['session.options']['cookie_path']     = $this['resources']->getUrl('root');
+        $this['session.options']['cookie_domain']   = $this['config']->get('general/cookies_domain');
+        $this['session.options']['cookie_secure']   = $this['config']->get('general/enforce_ssl');
     }
```
Comment (GawainLynch) [inline src/Application.php]: These are throwing problems in Travis:

```
PHP Notice:  Indirect modification of overloaded element of Bolt\Application has no effect in /home/travis/build/bolt/bolt/src/Application.php on line 92
PHP Stack trace:
PHP   1. {main}() /home/travis/build/bolt/bolt/tests/phpunit/nutty:0
PHP   2. Bolt\Tests\Nutty->getApp() /home/travis/build/bolt/bolt/tests/phpunit/nutty:21
PHP   3. Bolt\Tests\BoltUnitTest->getApp() /home/travis/build/bolt/bolt/tests/phpunit/nutty:12
PHP   4. Bolt\Application->initialize() /home/travis/build/bolt/bolt/tests/phpunit/unit/BoltUnitTest.php:44
PHP   5. Bolt\Application->initSession() /home/travis/build/bolt/bolt/src/Application.php:101
```

Comment (dwolfhub) [inline src/Application.php]: right, sorry. fixed.

**Comment by CarsonF** [comment]
@dwolfhub Yeah if you move it back into SessionSP I'll be good. I worked very hard in 3.0 dev cycle to shrink the Application class.

**Comment by dwolfhub** [comment]
@CarsonF Yea, I can do that, if you prefer. You had mentioned that you wanted to decouple the config from this service provider so i thought it made sense to make that bridge in the Application. Will patch.

**Comment by CarsonF** [comment]
Yeah sorry about flipping on you. Hard to know each step to take with a far away target.

**Comment by GawainLynch** [comment]
@CarsonF — if you're happy when @dwolfhub updates this, I am going to send it into `release/3.1`

**Comment by CarsonF** [comment]
There's still session options that will be missing now in most user's config.
I would just revert to what was there before and then add your line 90 after the defaults are set.

**Comment by CarsonF** [comment]
```php
$defaults = [
...
];
$app['session.options'] = $app['config']->get('general/session', []) + $defaults;
```

**Comment by CarsonF** [comment]
That still overwrites native php.ini options though...

**Comment by dwolfhub** [comment]
So the order in initConfig now goes:
1. defaults outlined on line 136
2. overwritten with values from php.ini
3. overwritten with values from session.options (if set either explicitly in a .bolt.php or by config.yml)

If we hard code anything in session.options it would overwrite php.ini values...

File: src/Provider/SessionServiceProvider.php [hunk 2]
```diff
@@ -105,17 +105,17 @@ public function boot(Application $app)
...
-            'name'            => 'bolt_session_',
-            'restrict_realm'  => true,
-            'save_handler'    => 'filesystem',
-            'save_path'       => 'cache://.sessions',
-            'cookie_lifetime' => $app['config']->get('general/cookies_lifetime'),
-            'cookie_path'     => $app['resources']->getUrl('root'),
-            'cookie_domain'   => $app['config']->get('general/cookies_domain'),
-            'cookie_secure'   => $app['config']->get('general/enforce_ssl'),
-            'cookie_httponly' => true,
-        ];
```
Comment (CarsonF) [inline src/Provider/SessionServiceProvider.php]:
```php
        $app['session.options'] = $app['config']->get('general/session', []) + [
            'name'            => 'bolt_session_',
            'restrict_realm'  => true,
            'cookie_lifetime' => $app['config']->get('general/cookies_lifetime'),
            'cookie_path'     => $app['resources']->getUrl('root'),
            'cookie_domain'   => $app['config']->get('general/cookies_domain'),
            'cookie_secure'   => $app['config']->get('general/enforce_ssl'),
            'cookie_httponly' => true,
        ];
```

**Comment by dwolfhub** [comment]
@CarsonF rebased and ready!

**Comment by CarsonF** [comment]
Looks good 👍

**Comment by GawainLynch** [comment]
Merged into `release/3.1` at https://github.com/bolt/bolt/commit/3e7cb6e904e4b1ac9b4d939d051f053568038cd7

**Pull request merged**